content-service/src/contentReplicator.js

```diff
@@ -1,273 +1,310 @@
+      const currentUserBlockNumber = await this._queryHighestBlockNumber(UserType)
+
+      // Query track and user with blocknumber ascending
+      // Sets the minimum blocknumber as the current highest blocknumber
+      // Limit the number of results to configured window
+      const tracks = await this.audiusLibs.Track.getTracks(
+        TrackIdWindowSize,
+        0,
+        null,
+        null,
+        'blocknumber:asc',
+        currentTrackBlockNumber + 1 /** minBlockNumber */
+      )
+      const users = await this.audiusLibs.User.getUsers(
+        UserIdWindowSize,
+        0,
+        null,
+        null,
+        null,
+        null,
+        currentUserBlockNumber + 1 /** minBlockNumber */
+      )
+
+      logger.info(`Beginning replication task for ${tracks.length} tracks, ${users.length} users.`)
+
+      // For each track and user, pin all associated multihashes
+      // This includes track segments, cover photos, profile pictures, and metadata
+      // Any record that has a higher block number is stored in the Blocks table for subsequent discovery provider queries
+      await Promise.all(
+        tracks.map(async track => this._replicateTrack(track)),
+        users.map(async user => this._replicateUser(user))
+      )
+      logger.info(`Replication task complete.`)
+    } catch (e) {
+      logger.error('REPLICATION ERROR', e)
+    } finally {
+      // if this does not get reset we will become stuck
+      this.replicating = false
+      // Query information and cleanup maximum usage of ipfs repo
+      await this._monitorDiskUsage()
     }
   }
 
-  async maxUsageExceeded () {
+  async _replicateTrack (track) {
     try {
-      let ipfsStats = await this.getIPFSRepoStats()
-      return ipfsStats.usagePercent > ipfsRepoMaxUsagePercent
+      const start = Date.now()
+      const trackId = track.track_id
+      const blocknumber = parseInt(track.blocknumber)
+      logger.info(`Replicating track with trackId ${trackId} at blocknumber ${blocknumber}.`)
+      let numMultihashes = 0
+
+      /** Pin all track metadata. */
+      const pinPromises = []
+      if (track.cover_art) {
+        pinPromises.push(this._replicateMultihash(track.cover_art, trackId, TrackType))
+        numMultihashes++
+      }
+      if (track.metadata_multihash) {
+        pinPromises.push(this._replicateMultihash(track.metadata_multihash, trackId, TrackType))
+        numMultihashes++
+      }
+      await Promise.all(pinPromises)
+
+      /** Pin all track segments. */
+      if (track.track_segments) {
+        let segments = track.track_segments
+        logger.info(`TrackID ${trackId} - ${segments.length} total segments`)
+        for (let i = 0; i < segments.length; i += TrackPinConcurrencyLimit) {
+          const slice = segments.slice(i, i + TrackPinConcurrencyLimit)
+          logger.info(`TrackID ${trackId} - Processing segments ${i} to ${i + TrackPinConcurrencyLimit}`)
+          await Promise.all(
```

Comment:
@hareeshnagaraj does the await get respected inside this for loop?

Reply:
yes it does actually

Reply:
@hareeshnagaraj oh i'm thinking of forEach 🤦‍♂